src/utils.ts

```diff
@@ -107,10 +107,56 @@ export async function getBalance(
   return value;
 }
 
-export function uiToNative(amount: number, decimals: number): BN {
-  return new BN(Math.round(amount * Math.pow(10, decimals)));
+export function nativeExp10(exponent: number) {
+  if (exponent < 0) {
+    throw new Error(
+      'BN does not support negative exponents (fractional values)'
+    );
+  }
+  return new BN('1' + '0'.repeat(exponent));
+}
+
+export function numberToFraction(value: number) {
```

Comment:
Tbh, I'm not sure I prefer theses supercomplicated functions compare to using a lib like `BigNumber` for theses kind of calculations

Reply:
I kinda agree with that, BN as a type is clunky and does not support everything that would be nice. But changing this is much more difficult to replace across the codebase than to have 1 or 2 complicated functions that we use transparently in the other codebases

Reply:
nono, just use `BigNumber` for this file calculations.

In and out of this file would be still `BN` and `BN` everywhere in the codebase